third_party/intel/lib/TritonGENToLLVM/Attributes.cpp

```diff
@@ -0,0 +1,436 @@
+      return cast<StringAttr>(ArrA1[0]) < cast<StringAttr>(ArrA2[0]);
+    return false;
+  };
+
+  llvm::sort(Vec.begin(), Vec.end(), Comp);
+  PassThroughAttr.setValue(ArrayAttr::get(&Ctx, Vec));
+
+  LLVM_DEBUG({
+    llvm::dbgs().indent(2) << AttributeList::PassThroughAttrName << ": ( ";
+    for (auto Item : Vec)
+      llvm::dbgs() << Item << " ";
+    llvm::dbgs() << ")\n";
+  });
+}
+
+static void addToPassThroughAttr(mlir::NamedAttribute &PassThroughAttr,
+                                 mlir::ArrayAttr NewAttrs, MLIRContext &Ctx) {
+  assert(PassThroughAttr.getName() == AttributeList::PassThroughAttrName &&
+         "PassThroughAttr is not valid");
+  assert(isa<ArrayAttr>(PassThroughAttr.getValue()) &&
+         "PassThroughAttr should have an ArrayAttr as value");
+
+  for (mlir::Attribute NewAttr : NewAttrs) {
+    if (auto ArrAttr = dyn_cast<ArrayAttr>(NewAttr)) {
+      assert(ArrAttr.size() == 2 && isa<StringAttr>(ArrAttr[0]));
+      NamedAttribute NamedAttr(cast<StringAttr>(ArrAttr[0]), ArrAttr[1]);
+      addToPassThroughAttr(PassThroughAttr, NamedAttr, Ctx);
+    } else if (auto StrAttr = dyn_cast<StringAttr>(NewAttr)) {
+      NamedAttribute NamedAttr(StrAttr, UnitAttr::get(&Ctx));
+      addToPassThroughAttr(PassThroughAttr, NamedAttr, Ctx);
+    } else
+      llvm_unreachable("Unexpected attribute kind");
+  }
+}
+
+//===----------------------------------------------------------------------===//
+// AttributeList Method Implementations
+//===----------------------------------------------------------------------===//
+
+AttributeList::AttributeList(const mlir::NamedAttrList &FnAttrs,
+                             const mlir::NamedAttrList &RetAttrs,
+                             llvm::ArrayRef<mlir::NamedAttrList> ParamAttrs)
+    : FnAttrs(FnAttrs), RetAttrs(RetAttrs), ParamAttrs(ParamAttrs) {}
+
+AttributeList &
+AttributeList::addAttrs(const AttrBuilder &FnAttrB, const AttrBuilder &RetAttrB,
+                        llvm::ArrayRef<mlir::NamedAttrList> Attrs) {
+  return addFnAttrs(FnAttrB).addRetAttrs(RetAttrB).addParamAttributes(Attrs);
+}
+
+AttributeList &AttributeList::addFnAttrs(const AttrBuilder &B) {
+  return addFnAttrs(B.getAttributes(), B.getContext());
+}
+
+AttributeList &AttributeList::addFnAttrs(const NamedAttrList &Attrs,
+                                         MLIRContext &Ctx) {
+  for (const NamedAttribute &NewFnAttr : Attrs) {
+    std::optional<NamedAttribute> ExistingFnAttr =
+        FnAttrs.getNamed(NewFnAttr.getName());
+    if (!ExistingFnAttr) {
+      FnAttrs.append(NewFnAttr);
+      continue;
+    }
+
+    // Merge the 'passthrough' attribute lists.
+    if (ExistingFnAttr->getName() == PassThroughAttrName) {
+      auto Attrs = cast<ArrayAttr>(NewFnAttr.getValue());
+      addToPassThroughAttr(*ExistingFnAttr, Attrs, Ctx);
+      FnAttrs.set(ExistingFnAttr->getName(), ExistingFnAttr->getValue());
+      continue;
+    }
+
+    llvm_unreachable("Function attribute already exists");
```

Comment:
This may be more readable.
```suggestion
    assert((!ExistingFnAttr || ExistingFnAttr->getName() == PassThroughAttrName) && "...");
    if (!ExistingFnAttr) {
      FnAttrs.append(NewFnAttr);
    } else if (ExistingFnAttr->getName() == PassThroughAttrName) {
      // Merge the 'passthrough' attribute lists.
      auto Attrs = cast<ArrayAttr>(NewFnAttr.getValue());
      addToPassThroughAttr(*ExistingFnAttr, Attrs, Ctx);
      FnAttrs.set(ExistingFnAttr->getName(), ExistingFnAttr->getValue());
    }
```